### ISSUE TITLE
Wrong Date type check for Arrow tables

### ISSUE THREAD

**Issue by mootari** [duckdb/duckdb-wasm#1231]
### What happens?

Dates with millisecond resolution in Arrow tables are misdetected as`date` instead of `date64[ms]`.

When an Arrow table is created via 
```js
arrow.tableFromJSON([{date: new Date()}])
```
the infered type is `Date` (typeId 8) with a `unit` of `1` ([source](https://github.com/apache/arrow/blob/45918a90a6ca1cf3fd67c256a7d6a240249e555a/js/src/type.ts#L319)). However, `arrowToSQLType()` checks against `DateMillisecond`(typeId -14) ([source](https://github.com/duckdb/duckdb-wasm/blob/master/packages/duckdb-wasm/src/json_typedef.ts#L30-L35)).

(It is unclear whether this is a bug in the Arrow implementation or DuckDB integration.)

### To Reproduce

observablehq.com example, via DuckDBClient:
~~~js
{
  const arrow = await import("https://cdn.observableusercontent.com/npm/apache-arrow@11.0.0/+esm");
  const table = arrow.tableFromJSON([{date: new Date("2023-02-01T12:34:56Z")}]);
  const type = table.schema.fields[0].type;
  const db = await DuckDBClient.of({t: table});
  const result = (await db.queryRow("select date d from t")).d.toISOString();
  return {result, type, typeId: type.typeId};
}
~~~

### Browser/Environment:

Chrome 111

### Device:

macOS 13.12.1

### DuckDB-Wasm Version:

1.24

### DuckDB-Wasm Deployment:

observablehq.com

### Full Name:

Fabian Iwand

### Affiliation:

Observable

**Comment by domoritz** [comment]
I'm happy to help with getting a fix into arrow if needed.

**Comment by carlopi** [comment]
It would be very welcome, I am not familiar yet with the Arrow conversion code and it's taking way too long to investigate this.

**Comment by domoritz** [comment]
I don't have time to investigate unfortunately but if there is a bug in Arrow, I can help with that.

**Comment by carlopi** [comment]
Thanks a lot! Will update with what I find.

**Comment by mootari** [comment]
According to the [enum inline docs](https://github.com/apache/arrow/blob/f4446c0f99b427a513f4a18d8ca3874fd5c56cee/js/src/enum.ts#L158-L176) any type values outside 0-17 are only used within TypeScript.

Am I reading that correctly to mean that the current behavior in Arrow works as designed, and that DuckDB would have to check the type's `unit` prop against `DateUnit.DAY` and `DateUnit.MILLISECOND`?

**Comment by dunstan** [comment]
Hi there. Any updates on this? I believe I just ran into the same issue.

**Comment by kylebarron** [comment]
With a small repro case [here](https://github.com/kylebarron/parquet-wasm/issues/413#issuecomment-1856379246), it look like this is a bug in Arrow JS

**Comment by domoritz** [comment]
Let's fix it upstream. I am happy to help get a fix merged. Who wants to make a pull request?

**Comment by domoritz** [comment]
I'm trying to investigate where the issue is. 

```ts
const date = new Date("2023-02-01T12:34:56Z");
console.log("original", date)

console.log("=> vec")
const vec = arrow.vectorFromArray([date])
console.log(vec.type)
console.log(vec.get(0)?.toISOString())

console.log("=> vec2")
const vec2 = arrow.vectorFromArray([date], new arrow.TimestampMillisecond)
console.log(vec2.type)
console.log(vec2.get(0))

console.log("=> table")
const table = arrow.tableFromJSON([{ date }])
console.log(table.schema.fields[0].type)
console.log(table.getChildAt(0)?.get(0).toISOString())

console.log("=> table2")
const table2 = arrow.tableFromIPC(arrow.tableToIPC(table));
console.log(table2.schema.fields[0].type)
console.log(table2.getChildAt(0)?.get(0).toISOString())
```

outputs (as of 96f686b81)

```
original 2023-02-01T12:34:56.000Z
=> vec
DateMillisecond {
  typeId: 8,
  unit: 1,
  toString: [Function: toString],
  ArrayType: [class Int32Array],
  [Symbol(Symbol.toStringTag)]: "Date",
  children: null,
  OffsetArrayType: [class Int32Array],
}
2023-02-01T12:34:56.000Z
=> vec2
TimestampMillisecond {
  typeId: 10,
  unit: 1,
  timezone: undefined,
  toString: [Function: toString],
  ArrayType: [class Int32Array],
  [Symbol(Symbol.toStringTag)]: "Timestamp",
  children: null,
  OffsetArrayType: [class Int32Array],
}
1675254896000
=> table
DateMillisecond {
  typeId: 8,
  unit: 1,
  toString: [Function: toString],
  ArrayType: [class Int32Array],
  [Symbol(Symbol.toStringTag)]: "Date",
  children: null,
  OffsetArrayType: [class Int32Array],
}
2023-02-01T12:34:56.000Z
=> table2
Date_ {
  typeId: 8,
  unit: 1,
  toString: [Function: toString],
  ArrayType: [class Int32Array],
  children: null,
  OffsetArrayType: [class Int32Array],
}
2023-02-01T12:34:56.000Z
```

It looks right to me but the issue seems to be related to serialization as well. 

Looking at the schema docs in https://github.com/apache/arrow/blob/main/format/Schema.fbs it seems like [DateMillisecond](https://github.com/apache/arrow/blob/96f686b81ba148f4d434846f0b9e161c538f131d/format/Schema.fbs#L245-L253) (date with unit milliseconds) is the right type since js dates are [timezone agnostic](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date). However, the [Rust docs](https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.html#variant.Date64) clarify that date should only encode the days and not the time. This seems like a compatibility type and we should discourage it similar to how rust did it recently: https://github.com/apache/arrow-rs/issues/5288. So yeah, I think we should use TimestampMillisecond with no timezone.

**Comment by domoritz** [comment]
Sent a fix in https://github.com/apache/arrow/pull/40892. Please take a look to see whether it now does what you expect.

**Comment by marvold-mw** [comment]
> Sent a fix in [apache/arrow#40892](https://github.com/apache/arrow/pull/40892). Please take a look to see whether it now does what you expect.

While I'm not the original reporter, your fix *does* do what *I* would expect when using JavaScript types with Arrow and (via Arrow) with DuckDB.  The TIMESTAMP type in DuckDB now comes back, which seems like the correct mapping for JS Date semantics.  And (as your unit tests confirm) Arrow converts them back to JavaScript Dates just fine.

Thanks for the fix, Dom!

**Comment by domoritz** [comment]
Fixed in https://github.com/apache/arrow/issues/40891.
